### PR TITLE
Natural convection Outflow bc

### LINKED ISSUE #24662: Adding correct temperature backflow treatment in Navier Stokes
## Reason
In natural convection problems with a pressure outlet and recirculation in this outlet, the backflow temperature in moose is taken as the extrapolated temperature from the cell nearest to the backflow boundary. This causes simulation divergence and unphysical temperatures in the domain.

Example problem - natural convection cooling of a fuel rod:
image

Current Navier Stoke's solution:
image
image

STAR-CCM+ reference solution:
image
image

## Design
A better treatment of the backflow temperature is needed in MOOSE, which allows the user to fix the backflow temperature in a pressure-outlet boundary condition.

## Impact
Add new capability to MOOSE.

### PR DESCRIPTION
**Pull request by tanoret**
refs #24662 closes #24662

New solutions with the BC improvement (left: Star-CCM+, right Pronghorn):

image

image

### CODE REVIEW

**Comment by GiudGiud** [comment]
you have 3 commits attached you should remove I think. They're from another PR

**Comment by tanoret** [comment]
> you have 3 commits attached you should remove I think. They're from another PR

Thanks. Should be good now

File: modules/navier_stokes/include/fvbcs/NSFVOutflowTemperatureBC.h [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  const Moose::Functor<ADReal> & _adv_quant;
+
+  /// x-velocity
+  const Moose::Functor<ADReal> & _u;
+  /// y-velocity
+  const Moose::Functor<ADReal> * const _v;
+  /// z-velocity
+  const Moose::Functor<ADReal> * const _w;
+
+  /// Density
```
Comment (GiudGiud) [inline modules/navier_stokes/include/fvbcs/NSFVOutflowTemperatureBC.h:37]:
```suggestion
  /// Backflow temperature
```

File: modules/navier_stokes/include/fvbcs/NSFVOutflowTemperatureBC.h [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "FVFluxBC.h"
+
+/**
+ * Constant velocity scalar advection boundary conditions
```
Comment (GiudGiud) [inline modules/navier_stokes/include/fvbcs/NSFVOutflowTemperatureBC.h:15]:
```suggestion
 * Temperature advection boundary condition allowing for inflow and outflow
```

File: modules/navier_stokes/test/tests/finite_volume/ins/natural_convection/fuel_cavity.i [hunk 3]
```diff
@@ -0,0 +1,383 @@
+# ========================================================================
+# The purpose of this MOOSE scripts is to solve a 2-D axisymmetric
+#     roblem with the following details:
```
Comment (GiudGiud) [inline modules/navier_stokes/test/tests/finite_volume/ins/natural_convection/fuel_cavity.i:3]:
```suggestion
#     problem with the following details:
```

File: modules/navier_stokes/test/tests/finite_volume/ins/natural_convection/fuel_cavity.i [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+    expression = 'g * beta * (T - T_init) * pow(width, 3) / (nu*alpha) + 1.0'
+    block = 0
+  []
+  [htc]
+    type = ParsedAux
+    variable = htc
+    coupled_variables = 'Ra'
+    constant_names = 'Pr'
+    constant_expressions = '${fparse cp*mu/k}'
+    #expression = '0.1 * Ra * Pr'
```
Comment (GiudGiud) [inline modules/navier_stokes/test/tests/finite_volume/ins/natural_convection/fuel_cavity.i:321]:
```suggestion
```

**Review by GiudGiud — CHANGES_REQUESTED** [review]

File: modules/navier_stokes/include/fvbcs/NSFVOutflowTemperatureBC.h (hunk 1 above)
Comment (tanoret) [reply modules/navier_stokes/include/fvbcs/NSFVOutflowTemperatureBC.h]: Tx!

File: modules/navier_stokes/include/fvbcs/NSFVOutflowTemperatureBC.h [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+{
+public:
+  NSFVOutflowTemperatureBC(const InputParameters & parameters);
+
+  static InputParameters validParams();
+
+protected:
+  virtual ADReal computeQpResidual() override;
+
+  /// Density
```
[the remark this answers is not in the document]
Comment (tanoret) [reply modules/navier_stokes/include/fvbcs/NSFVOutflowTemperatureBC.h]: Tx!

File: modules/navier_stokes/include/fvbcs/NSFVOutflowTemperatureBC.h (hunk 2 above)
Comment (tanoret) [reply modules/navier_stokes/include/fvbcs/NSFVOutflowTemperatureBC.h]: Tx!

File: modules/navier_stokes/test/tests/finite_volume/ins/natural_convection/fuel_cavity.i (hunk 3 above)
Comment (tanoret) [reply modules/navier_stokes/test/tests/finite_volume/ins/natural_convection/fuel_cavity.i]: Tx!

File: modules/navier_stokes/test/tests/finite_volume/ins/natural_convection/tests [hunk 6]
```diff
@@ -1,10 +1,25 @@
...
+    requirement = 'The system shall be able to predict correct temperature profiles in natrual convection cooling of fuel cavities.'
+  []
+  [natural_circulation_fuel_cavity_csv]
+    type = 'CSVDiff'
+    input = 'fuel_cavity.i'
+    csvdiff = 'fuel_cavity_out.csv'
+    requirement = 'The system shall be able to correctly predict the maximum fluid and solid temperature during solidification.'
+    rel_err = 1.0E-7
+    recover = false
+  []
```
[the remark this answers is not in the document]
Comment (tanoret) [reply modules/navier_stokes/test/tests/finite_volume/ins/natural_convection/tests]: Thanks. Retaines the csv test only

**Review by GiudGiud — DISMISSED** [review]
lgtm

@lindsayad want to take a look?

File: modules/navier_stokes/src/fvbcs/NSFVOutflowTemperatureBC.C [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    return rho_cp_face * _var.getBoundaryFaceValue(*_face_info, determineState()) * vol_flux;
+  }
+  else
+  {
+    auto face_arg =
+        makeFace(*_face_info, Moose::FV::LimiterType::Upwind, MetaPhysicL::raw_value(vol_flux) < 0);
+    auto rho_cp_face = _adv_quant(face_arg, determineState());
+    auto backflow_T = _backflow_T(face_arg, determineState());
+    return rho_cp_face * backflow_T * vol_flux;
+  }
```
Comment (lindsayad) [inline modules/navier_stokes/src/fvbcs/NSFVOutflowTemperatureBC.C:74]: So the default functor is rho * cp * T. And then you are multiplying by the variable value, which is the temperature. So by default you will have the square of temperature. Also when composing `face_arg` I think we should think carefully about how we want to compose it. I would kind of think we should
1. just use `boundary_face`
2. think harder about what we supply for the third argument to `makeFace`. The part in the `if` branch is fine ... that should upwind any variable values used in the evaluation of the `_adv_quant` functor. But in the `else` branch we are being implicitly asked to provide "upwind" values for `_adv_quant` when we are outside the bounds of the domain

File: modules/navier_stokes/test/tests/finite_volume/ins/natural_convection/fuel_cavity.i [hunk 8]
```diff
@@ -0,0 +1,382 @@
...
+  [fluid_side]
+    type = BreakBoundaryOnSubdomainGenerator
+    input = 'interface'
+    boundaries = 'top bottom'
+  []
+[]
+
+[GlobalParams]
+  # retain behavior at time of test creation
+  two_term_boundary_expansion = false
```
Comment (lindsayad) [inline modules/navier_stokes/test/tests/finite_volume/ins/natural_convection/fuel_cavity.i:67]:
```suggestion
```

File: modules/navier_stokes/test/tests/finite_volume/ins/natural_convection/fuel_cavity.i [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+    block = 1
+    initial_condition = ${T_init}
+    scaling = 1e-3
+  []
+[]
+
+[FVKernels]
+
+  # inactive = 'u_time v_time temp_time Ts_time'
+
```
Comment (lindsayad) [inline modules/navier_stokes/test/tests/finite_volume/ins/natural_convection/fuel_cavity.i:115]:
```suggestion
```

File: modules/navier_stokes/test/tests/finite_volume/ins/natural_convection/tests [hunk 10]
```diff
@@ -1,10 +1,18 @@
...
     exodiff = 'natural_circulation_dogleg_out.e'
     requirement = 'The system shall be able to use pressure inlet and outlet boundary conditions to compute open, chimney-type natural circulation problems.'
   []
+  [natural_circulation_fuel_cavity_csv]
+    type = 'CSVDiff'
+    input = 'fuel_cavity.i'
+    csvdiff = 'fuel_cavity_out.csv'
+    requirement = 'The system shall be able to correctly predict the maximum fluid and solid temperatures.'
+    rel_err = 1.0E-7
+    recover = false
```
Comment (lindsayad) [inline modules/navier_stokes/test/tests/finite_volume/ins/natural_convection/tests:16]: do we need this?

**Review by lindsayad — CHANGES_REQUESTED** [review]
Thank you so much for the contribution @tanoret. This is a great addition

File: modules/navier_stokes/src/fvbcs/NSFVOutflowTemperatureBC.C (hunk 7 above)
Comment (tanoret) [reply modules/navier_stokes/src/fvbcs/NSFVOutflowTemperatureBC.C]: Thanks. Changed to `boundary_face`. Results didn't change for the test

File: modules/navier_stokes/test/tests/finite_volume/ins/natural_convection/fuel_cavity.i (hunk 8 above)
Comment (tanoret) [reply modules/navier_stokes/test/tests/finite_volume/ins/natural_convection/fuel_cavity.i]: Tx! Removed

File: modules/navier_stokes/test/tests/finite_volume/ins/natural_convection/fuel_cavity.i (hunk 9 above)
Comment (tanoret) [reply modules/navier_stokes/test/tests/finite_volume/ins/natural_convection/fuel_cavity.i]: Tx! Removed

File: modules/navier_stokes/test/tests/finite_volume/ins/natural_convection/tests (hunk 10 above)
Comment (tanoret) [reply modules/navier_stokes/test/tests/finite_volume/ins/natural_convection/tests]: Thanks! Removed `recover`

File: modules/navier_stokes/src/fvbcs/NSFVOutflowTemperatureBC.C (hunk 7 above)
Comment (tanoret) [reply modules/navier_stokes/src/fvbcs/NSFVOutflowTemperatureBC.C]: Also changed the 'rho * cp * T' functor for 'rho' and 'cp'

File: modules/navier_stokes/test/tests/finite_volume/ins/natural_convection/tests [hunk 11]
```diff
@@ -1,10 +1,17 @@
...
     exodiff = 'natural_circulation_dogleg_out.e'
     requirement = 'The system shall be able to use pressure inlet and outlet boundary conditions to compute open, chimney-type natural circulation problems.'
   []
+  [natural_circulation_fuel_cavity_csv]
+    type = 'CSVDiff'
+    input = 'fuel_cavity.i'
+    csvdiff = 'fuel_cavity_out.csv'
+    requirement = 'The system shall be able to correctly predict the maximum fluid and solid temperatures.'
+    rel_err = 1.0E-7
+  []
```
Comment (GiudGiud) [inline modules/navier_stokes/test/tests/finite_volume/ins/natural_convection/tests:16]:
```suggestion
    method = opt
  []
```
too slow in debug

File: modules/navier_stokes/src/fvbcs/NSFVOutflowTemperatureBC.C [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+registerMooseObject("NavierStokesApp", NSFVOutflowTemperatureBC);
+
+InputParameters
+NSFVOutflowTemperatureBC::validParams()
+{
+  InputParameters params = FVFluxBC::validParams();
+  params.addClassDescription("Outflow velocity temperature advection boundary conditions for "
+                             "finite volume method allowing for thermal backflow.");
+  params.addParam<MooseFunctorName>(NS::density, "The name of the density");
+  params.addParam<MooseFunctorName>(NS::cp, "The name of the specific heat");
```
Comment (lindsayad) [inline modules/navier_stokes/src/fvbcs/NSFVOutflowTemperatureBC.C:23]:
```suggestion
  params.addRequiredParam<MooseFunctorName>(NS::density, "The name of the density");
  params.addRequiredParam<MooseFunctorName>(NS::cp, "The name of the specific heat");
```

File: modules/navier_stokes/src/fvbcs/NSFVOutflowTemperatureBC.C [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+  const auto state = determineState();
+
+  ADRealVectorValue v(_u(boundary_face, state));
+  if (_v)
+    v(1) = (*_v)(boundary_face, state);
+  if (_w)
+    v(2) = (*_w)(boundary_face, state);
+
+  auto vol_flux = v * _normal;
+  auto rho_cp_face = _rho(boundary_face, state) * _cp(boundary_face, state);
```
Comment (lindsayad) [inline modules/navier_stokes/src/fvbcs/NSFVOutflowTemperatureBC.C:62]:
```suggestion
  const auto vol_flux = v * _normal;
  const auto rho_cp_face = _rho(boundary_face, state) * _cp(boundary_face, state);
```

File: modules/navier_stokes/src/fvbcs/NSFVOutflowTemperatureBC.C [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+
+  if (vol_flux > 0)
+  {
+    return rho_cp_face * _var(boundary_face, state) * vol_flux;
+  }
```
Comment (lindsayad) [inline modules/navier_stokes/src/fvbcs/NSFVOutflowTemperatureBC.C:67]:
```suggestion
    return rho_cp_face * _var(boundary_face, state) * vol_flux;
```

**Review by lindsayad — CHANGES_REQUESTED** [review]
I think after these changes we are good to go on my end

File: modules/navier_stokes/src/fvbcs/NSFVOutflowTemperatureBC.C (hunk 12 above)
Comment (tanoret) [reply modules/navier_stokes/src/fvbcs/NSFVOutflowTemperatureBC.C]: Thanks. Changed

File: modules/navier_stokes/src/fvbcs/NSFVOutflowTemperatureBC.C (hunk 13 above)
Comment (tanoret) [reply modules/navier_stokes/src/fvbcs/NSFVOutflowTemperatureBC.C]: Thanks. Changed

File: modules/navier_stokes/src/fvbcs/NSFVOutflowTemperatureBC.C (hunk 14 above)
Comment (tanoret) [reply modules/navier_stokes/src/fvbcs/NSFVOutflowTemperatureBC.C]: Thanks. Chaged

**Review by lindsayad — APPROVED** [review]

**Comment by lindsayad** [comment]
Much appreciated @tanoret. You're becoming a regular MOOSE developer!
